### DISCUSSION TITLE
What do you think about my Typography extension?

### DISCUSSION THREAD

**Discussion by GiovanniSalmeri** [datenstrom/community#741] [Ask a question] [answered]
Hello. A new little extension [Typography](https://github.com/GiovanniSalmeri/yellow-typography). This extension offers three "typographic" enhancements: I have put them together in one extension (after some hesitation) because each of them require little code (other than external JS libraries) and it seems to me probable that all of them are useful for the same scenarios:

1. possibility of styling the title of the page with markdown
2. so-called "smart quotes" and similar characters
3. hyphenation (also in browser that do not natively support it)

As regards 2: I am not a huge fan of similar systems of automatic conversion, above all because there are two superior systems: directly writing the right Unicode characters (e.g. the .txt format in Project Gutemberg does so), or inserting (for the quotes) the tag `<q>`. But each of those systems can be inconvenient: for the first, text editors in general have a bad (or inexistent) support for those characters; for the second, one would be forced to use HTML, since no dialect of markdown has support for `<q>`. Moreover, the first and most known system for automatic conversion of quotes and similar, Smartypants, was written by Gruber (and translated in PHP by Fortin) as a companion for Markdown, and also Commonmark has something very similar. In this implementation I tried to maintain the code as simple and fast as possible, and above all I made it language-sensitive: quotes are always written `"` and `'` (those are the ASCII proper characters), but they are converted differently according to the language of the page. Moreover, spaces receive a particular treatment in French texts. Obviously, the heuristic can sometimes fail, as in all similar systems.

As regards 3: the native support for hyphenations is getting much better in the latest months, so the aim of this function is essentially to support older or niche browsers.

Edit: I have just discovered that a Smartypants extension did exist: https://github.com/datenstrom/yellow/discussions/7#discussioncomment-246789

**Comment by markseuffert** [comment, marked as answer]
As regards to 1: Perhaps it's possible to style the title of a page without the need for a custom layout. At the moment the Typography extension requires to customise the default layout. Which is not that easy for new users. The extension could use the event `onParseContentHtml()`, find the HTML element for the title, then replace it with the typographical enhanced title. Not sure what the disadvantages are, on one side it feels wrong to do expensive post-processing of HTML, on the other side this would make life easier.

Just a technical idea... stumbled upon it while commenting #998.

**Reply by GiovanniSalmeri** [reply]
Thank you for the good idea, I will try it as soon as possible! (Perhaps I didn't think about it because [this answer](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454) left on me a lasting trauma 😁)

**Reply by markseuffert** [reply]
As a multi model thinker you are allowed to choose the right tool for the right task. 😄
